messages: Add support for passing user IDs for pm-with clause.

priyank-p opened the pull request:
I normally don't work in Python so there could be some obvious things wrong! @eeshangarg you might want to have a look at this given your previous work on this area.

I currently trying to work on making the frontend use this syntax but haven't got time to do so and it's not as easy to trace back all the `narrow_state` and `Filter` code.

**Testing Plan:** <!-- How have you tested? -->
- Added tests
- Verified that narrow view works as expected in development environment.

timabbott commented on zerver/views/messages.py:
```diff
@@ -336,7 +344,7 @@ def by_pm_with(self, query: Query, operand: str, maybe_negate: ConditionTransfor
                 narrow_profile = get_user_including_cross_realm(target_email, self.user_realm)
             except UserProfile.DoesNotExist:
-                raise BadNarrowOperator('unknown user ' + operand)
+                raise BadNarrowOperator('unknown user in operand')
```
What's the story with these changes?  It's generally better to provide the object with the error, as we had before.

priyank-p replied:
I was getting this error when linting:
```
mypy      | zerver/views/messages.py:328: error: Unsupported operand types for + ("str" and "Iterable[int]")
mypy      | zerver/views/messages.py:328: note: Right operand is of type "Union[str, Iterable[int]]"
```

I figured out though that using `str(operand)` should fix this.

timabbott commented on zerver/views/messages.py:
```diff
@@ -300,13 +300,21 @@ def by_id(self, query: Query, operand: str, maybe_negate: ConditionTransform) ->
+                my_email=self.user_profile.email,
+            )
+        else:
+            emails = raw_pm_with_emails_by_ids(
```
Let's add a comment here:
"""
This is where we handle passing a list of user IDs for the narrow, which is the 
preferred/cleaner API.  Ideally, we'd refactor the below to natively use user IDs, 
rather than having mapping user IDs to emails here.
"""

timabbott commented:
This generally looks great.  I posted a few comments; can you also update https://zulipchat.com/api/construct-narrow (or really, the "search" article it links to) to make clear we now support passing user IDs in the narrow?  

I think ideally we'd restructure the `by_pm_with` function to have a cleaner structure (in general, going through IDs is much better than going through email addresses), but after fiddling a bit I can see it's probably not going to be a super simple refactor to do so, and I think it's reasonable to open a follow-up issue for that as it's not on the critical path to doing this feature.

priyank-p commented:
Updated this to: pass in the bad operand with the error, add the comment mentioned, and add the documentation. I also added one more commit for adding json syntax highlighting to `/api/construct-narrow` page

timabbott commented:
This is great, merged, thanks @priyank-p!  I made one tweak, which was to move the doc update to `/api/construct-narrow`, since I felt listing the user ID there did make the user-facing search docs worse.

timabbott commented:
For follow-ups:
* You can do similar adjustments for what the "sender:" syntax supports.
* You can migrate the frontend to use the new API
* Looking at how this looks on /api/construct-narrow, we might want to add an example of the new syntax for group PMs of doing a list ... I'm actually not confident the docs we just added are correct on this front for the /api context of actual API clients...
* I opened https://github.com/zulip/zulip-mobile/issues/3525 to notify the mobile team.
